src/Processors/Transforms/Streaming/WatermarkStamper.cpp

```diff
@@ -146,33 +146,107 @@ void WatermarkStamper::preProcess(const Block & header)
         initTimeoutTimer(params.timeout_interval);
 }
 
-void WatermarkStamper::process(Chunk & chunk)
+template <bool apply_watermark_per_row>
+ALWAYS_INLINE Int64 WatermarkStamper::calculateWatermark(Int64 event_ts) const
 {
+    if (params.delay_interval)
+    {
+        auto event_ts_bias = addTime(
+            event_ts,
+            params.delay_interval.unit,
+            -1 * params.delay_interval.interval,
+            *params.window_params->time_zone,
+            params.window_params->time_scale);
+
+        if constexpr (apply_watermark_per_row)
+            return event_ts_bias;
+        else
+            return calculateWatermarkBasedOnWindowImpl(event_ts_bias);
+    }
+    else
+    {
+        if constexpr (apply_watermark_per_row)
+            return event_ts;
+        else
+            return calculateWatermarkBasedOnWindowImpl(event_ts);
+    }
+}
+
+void WatermarkStamper::processAfterUnmuted(Chunk & chunk)
+{
+    assert(!chunk.hasRows() && chunk.isHistoricalDataEnd());
+
     switch (params.mode)
     {
         case WatermarkStamperParams::EmitMode::PERIODIC: {
             processPeriodic(chunk);
             break;
         }
         case WatermarkStamperParams::EmitMode::WATERMARK: {
-            assert(params.window_params);
-            if (params.window_params->time_col_is_datetime64)
-                processWatermark<ColumnDateTime64, false>(chunk);
-            else
-                processWatermark<ColumnDateTime, false>(chunk);
+            auto muted_watermark_ts = calculateWatermark<false>(max_event_ts);
+            if (muted_watermark_ts != INVALID_WATERMARK)
+                chunk.setWatermark(muted_watermark_ts);
             break;
         }
         case WatermarkStamperParams::EmitMode::WATERMARK_PER_ROW: {
-            assert(params.window_params);
-            if (params.window_params->time_col_is_datetime64)
-                processWatermark<ColumnDateTime64, true>(chunk);
-            else
-                processWatermark<ColumnDateTime, true>(chunk);
+            auto muted_watermark_ts = calculateWatermark<true>(max_event_ts);
+            if (muted_watermark_ts != INVALID_WATERMARK)
+                chunk.setWatermark(muted_watermark_ts);
             break;
         }
         default:
             break;
     }
+}
+
+template <bool mute_watermark>
+void WatermarkStamper::process(Chunk & chunk)
```

Comment:
Let's separate this function into 2 functions instead of merging them in one since they are completely different logic without sharing.

Reply:
`processWithMutedWatermark(...)`
`process(...)

Reply:
Let's make our code simple and easy to read / understand.